File: backend/app/services/ai/agent_service.py
import json
import logging
from typing import Dict, Any, List, Optional
from sqlalchemy.orm import Session
from ...db.session import SessionLocal
from ...models.chapter import Chapter
from ...models.character import Character
from ...models.scene import Scene
from ...schemas.agent import AgentRequest, AgentResponse, ToolCall
from .gemini_provider import GeminiProvider
from ...core.config import settings
from ...services.llm import LLMService 

logger = logging.getLogger(__name__)

class AgentService:
    def __init__(self):
        self.llm = GeminiProvider(api_key=settings.GEMINI_API_KEY)

    def process_request(self, request: AgentRequest) -> AgentResponse:
        db = SessionLocal()
        try:
            # 1. Build Context
            context_str = self._build_context(db, request)
            
            # 2. Build System Prompt
            language = request.context.language or "zh"
            system_prompt = self._build_system_prompt(context_str, language)
            
            # 3. Construct Message History
            messages_str = ""
            for msg in request.history:
                role = "User" if msg["role"] == "user" else "Agent"
                messages_str += f"{role}: {msg['content']}\n"
            messages_str += f"User: {request.message}\nAgent:"

            # 4. Call LLM
            full_prompt = f"{messages_str}"
            # Note: GeminiProvider.generate_text supports system_instruction
            raw_response = self.llm.generate_text(full_prompt, system_instruction=system_prompt)
            
            # 5. Parse Response
            agent_response = self._parse_response(raw_response)
            
            # 6. Execute Tool (if any)
            if agent_response.action:
                tool_result = self._execute_tool(db, agent_response.action, request)
                if tool_result:
                    agent_response.response += f"\n\n[Tool Output]: {tool_result}"
            
            return agent_response

        except Exception as e:
            logger.error(f"Agent processing failed: {e}")
            return AgentResponse(
                thought="System Error",
                response="I encountered an internal error while processing your request.",
                action=None
            )
        finally:
            db.close()

    def _build_context(self, db: Session, request: AgentRequest) -> str:
        ctx = request.context
        context_parts = []
        
        if ctx.project_id:
            chars = db.query(Character).filter(Character.project_id == ctx.project_id).all()
            char_list = ", ".join([c.name for c in chars])
            context_parts.append(f"Project Characters: {char_list}")

        if ctx.chapter_id:
            chapter = db.query(Chapter).filter(Chapter.id == ctx.chapter_id).first()
            if chapter:
                context_parts.append(f"Current Chapter: {chapter.title}")
                # Add summary or excerpt if needed, but keep it brief for now
                if chapter.summary:
                     context_parts.append(f"Chapter Summary: {chapter.summary}")
                if chapter.content:
                     # Provide first 500 chars as glimpse
                     snippet = chapter.content[:500] + "..." if len(chapter.content) > 500 else chapter.content
                     context_parts.append(f"Chapter Start: {snippet}")

        return "\n".join(context_parts)

    def _build_system_prompt(self, context_str: str, language: str = "zh") -> str:
        lang_instruction = "You MUST respond in CHINESE (Simplified)." if language == "zh" else "You MUST respond in ENGLISH."
        return f"""You are the 'Director's Assistant' in NovaStory, an AI-powered creative writing and directing platform.
Your goal is to assist the user in directing their story, managing characters, and visualizing scenes.

LANGUAGE:
{lang_instruction}

CONTEXT:
{context_str}

AVAILABLE TOOLS:
1. analyze_chapter(chapter_id: str) - Analyze the current chapter for mood, pacing, and characters.
2. generate_timeline(chapter_id: str) - Automatically break down the chapter into scenes (storyboard).
3. get_character_info(name: str) - Get details about a specific character.
4. update_character_info(name: str, description: str, visual_tags: dict) - Update a character's traits.

RESPONSE FORMAT:
You must output a JSON object with the following structure:
{{
  "thought": "Your internal reasoning process...",
  "response": "Your natural language response to the user...",
  "action": {{ "tool_name": "...", "arguments": {{...}}, "reason": "..." }} (OPTIONAL)
}}

Do not include markdown formatting (like ```json) in your output. Just the raw JSON.
If no tool is needed, set "action" to null.
"""

    def _parse_response(self, raw_text: str) -> AgentResponse:
        try:
            # Clean markdown if present
            clean_text = raw_text.replace("```json", "").replace("```", "").strip()
            data = json.loads(clean_text)
            return AgentResponse(**data)
        except json.JSONDecodeError:
            # Fallback for non-JSON responses (Self-Healing attempt could go here)
            return AgentResponse(
                thought="Failed to parse JSON",
                response=raw_text, # Assume the whole text is the response
                action=None
            )

    def _execute_tool(self, db: Session, action: ToolCall, request: AgentRequest) -> str:
        try:
            if action.tool_name == "analyze_chapter":
                chapter_id = action.arguments.get("chapter_id") or request.context.chapter_id
                if not chapter_id: return "Error: No chapter ID provided."
                
                chapter = db.query(Chapter).filter(Chapter.id == chapter_id).first()
                if not chapter: return "Error: Chapter not found."
                
                # Reuse LLMService analysis
                analysis = LLMService.analyze_content(chapter.content or "")
                return f"Analysis Result: {json.dumps(analysis, indent=2)}"

            elif action.tool_name == "generate_timeline":
                chapter_id = action.arguments.get("chapter_id") or request.context.chapter_id
                if not chapter_id: return "Error: No chapter ID provided."
                
                # We can trigger the actual generation or just return a suggestion
                # Here we'll return a message that it's ready to be generated via UI, 
                # or actually call the generation service if we want to bypass UI state (risky for sync).
                # Better to just instruct the user or return a plan.
                # However, the user asked for "Automatic storyboard", so let's generate the data.
                
                chapter = db.query(Chapter).filter(Chapter.id == chapter_id).first()
                if not chapter or not chapter.content: return "Error: Chapter content empty."
                
                timeline_data = LLMService.generate_timeline(chapter.content)
                
                # Optionally save to DB here? 
                # DirectorMode.tsx pulls from DB. So we should save it.
                # But that might overwrite existing work.
                # For now, let's return the plan and let the user confirm in chat? 
                # Or just save it. Let's save it for "Agentic" feel.
                
                # Clear old scenes?
                # db.query(Scene).filter(Scene.chapter_id == chapter_id).delete()
                
                # Add new scenes
                saved_count = 0
                for i, scene_data in enumerate(timeline_data):
                     scene = Scene(
                         chapter_id=chapter_id,
                         index=i,
                         visual_prompt=scene_data.get("visual_prompt", ""),
                         audio_prompt=scene_data.get("audio_prompt", ""),
                         duration=scene_data.get("duration", 3.0),
                         shot_type=scene_data.get("camera", {}).get("shot_type", "medium"),
                         camera_movement=scene_data.get("camera", {}).get("movement", "static"),
                         camera_angle=scene_data.get("camera", {}).get("angle", "eye_level")
                     )
                     db.add(scene)
                     saved_count += 1
                
                db.commit()
                return f"Successfully generated and saved {saved_count} scenes for chapter '{chapter.title}'."

            elif action.tool_name == "get_character_info":
                name = action.arguments.get("name")
                char = db.query(Character).filter(Character.name == name, Character.project_id == request.context.project_id).first()
                if char:
                    return f"Name: {char.name}\nRole: {char.role}\nDescription: {char.description}\nTags: {char.visual_tags}"
                return "Character not found."

            elif action.tool_name == "update_character_info":
                name = action.arguments.get("name")
                desc = action.arguments.get("description")
                tags = action.arguments.get("visual_tags")
                
                char = db.query(Character).filter(Character.name == name, Character.project_id == request.context.project_id).first()
                if char:
                    if desc: char.description = desc
                    if tags: char.visual_tags = tags
                    db.commit()
                    return f"Updated character '{name}'."
                return "Character not found."

            return "Unknown tool."
        
        except Exception as e:
            logger.error(f"Tool execution failed: {e}")
            return f"Tool execution failed: {str(e)}"

File: backend/src/services/ai/agent_service.ts
import { AgentRequest, AgentResponse } from '../../schemas/agent';
import {
  needsConfirmation,
  type AgentOsDecision,
} from '../../schemas/agent_os';
import { logger } from '../../core/logging';
import { LLMService } from '../llm';
import { WritingService } from './writing_service';
import {
  decisionFromRoute,
  freeTextAnswerFallback,
  resolveAgentRoute,
} from './agent_route';
import { AgentExecutor } from './agent_executor';

function stripThink(text: string): string {
  return text.replace(/<think>[\s\S]*?<\/think>/g, '').trim();
}

export class AgentService {
  async processRequest(request: AgentRequest): Promise<AgentResponse> {
    try {
      const projectId = request.context.project_id;
      if (!projectId) {
        return this.projectlessChat(request);
      }

      const decision = await this.decide(request, projectId);
      let actions = decision.actions || [];
      const confirm = needsConfirmation(actions);

      // Auto-run read-only actions so the user gets answers immediately
      let autoNotes: string[] = [];
      let autoResults: any[] = [];
      if (!confirm && actions.length > 0) {
        // ANSWER_QUESTION with placeholder: expand via free-text if answer is just the user message
        actions = await this.hydrateAnswerQuestions(actions, request);

        const results = await AgentExecutor.executeAll(actions, {
          projectId,
          chapterId: request.context.chapter_id,
          language: request.context.language,
          apply: true,
        });
        autoResults = results;
        autoNotes = results.map(
          (r) =>
            `[${r.op}] ${r.status}${r.message ? ': ' + r.message : ''}`
        );
      }

      // Build user-facing response
      let responseText = decision.response || '';
      if (!responseText) {
        const answer = actions.find((a: any) => a.op === 'ANSWER_QUESTION') as
          | { answer?: string }
          | undefined;
        responseText = answer?.answer || (actions.length
          ? `计划执行 ${actions.length} 步操作，请确认。`
          : '好的。');
      }
      if (autoNotes.length) {
        // Keep notes short for character analysis (card shows detail)
        const hasRichCard = autoResults.some(
          (r) =>
            r.op === 'ANALYZE_CHAPTER_CHARACTERS'
            || r.op === 'ANALYZE_CHAPTER'
            || r.op === 'RUN_CONSISTENCY_CHECK'
        );
        if (!hasRichCard) {
          responseText += '\n\n' + autoNotes.join('\n');
        }
      }

      const firstMutating = actions.find((a: any) => needsConfirmation([a]));

      return {
        thought: decision.thought,
        response: responseText,
        actions: actions as any[],
        results: autoResults,
        needs_confirmation: confirm && actions.length > 0,
        action: firstMutating
          ? {
              tool_name: (firstMutating as any).op,
              arguments: firstMutating as any,
              reason: decision.thought,
            }
          : null,
      };
    } catch (error) {
      logger.error(`Agent processing failed: ${error}`);
      return {
        thought: 'System Error',
        response:
          '处理请求时发生内部错误。请稍后重试，或检查本地 LLM 是否已启动。',
        actions: [],
        results: [],
        needs_confirmation: false,
      };
    }
  }

  /**
   * When route maps to ANSWER_QUESTION with focus==userMessage, generate a real reply.
   */
  private async hydrateAnswerQuestions(
    actions: any[],
    request: AgentRequest
  ): Promise<any[]> {
    const out = [];
    for (const a of actions) {
      if (a.op !== 'ANSWER_QUESTION') {
        out.push(a);
        continue;
      }
      const ans = String(a.answer || '').trim();
      const msg = String(request.message || '').trim();
      if (ans && ans !== msg && ans.length > msg.length + 20) {
        out.push(a);
        continue;
      }
      try {
        const provider = LLMService.getProvider();
        const text = await provider.generateText(
          msg,
          '你是 NovaStory 写作助手。用简体中文简洁回答用户关于剧本/章节/角色的问题。不要假装已修改数据库。'
        );
        out.push({ op: 'ANSWER_QUESTION', answer: stripThink(text) || ans || msg });
      } catch {
        out.push(a);
      }
    }
    return out;
  }

  private async projectlessChat(request: AgentRequest): Promise<AgentResponse> {
    const provider = LLMService.getProvider();
    const lang =
      request.context.language === 'en' ? 'English' : 'Simplified Chinese';
    const text = await provider.generateText(
      request.message,
      `You are NovaStory assistant. Reply in ${lang}. No tools available without a project context.`
    );
    return {
      thought: 'No project context',
      response: stripThink(text),
      actions: [],
      results: [],
      needs_confirmation: false,
    };
  }

  /**
   * P0: preferred_op / keyword shortcut / strict mini Route Schema / free-text fallback.
   * No loose actions[] planner.
   */
  private async decide(
    request: AgentRequest,
    projectId: number
  ): Promise<AgentOsDecision> {
    const preferredOp =
      (request.context as any)?.preferred_op
      || (request.context as any)?.preferredOp
      || null;

    let chapterTitle: string | null = null;
    let overrides: Partial<Record<string, string>> | null = null;
    try {
      const bundle = await WritingService.loadBundleForAgent(
        projectId,
        request.context.chapter_id
      );
      const active =
        bundle.chapters.find((c) => c.id === bundle.activeId) ||
        bundle.chapters[0];
      chapterTitle = active?.title || null;
      overrides =
        (bundle.settings?.agent_prompts_override as Partial<
          Record<string, string>
        > | null) || null;
    } catch {
      /* ignore */
    }

    const history = (request.history || [])
      .slice(-4)
      .map((m: any) => {
        const role = m.role === 'user' ? 'U' : 'A';
        return `${role}:${String(m.content || '').slice(0, 120)}`;
      })
      .join('\n');

    const resolved = await resolveAgentRoute({
      userMessage: request.message,
      chapterId: request.context.chapter_id,
      chapterTitle,
      routeHint: request.context.route,
      preferredOp,
      historyTail: history,
      overrides: overrides as any,
    });

    if (resolved) {
      logger.info(
        `Agent route OK source=${resolved.source} intent=${resolved.route.intent}`
      );
      return decisionFromRoute(resolved.route, {
        chapterId: request.context.chapter_id,
        userMessage: request.message,
      });
    }

    logger.warn('Agent route failed — free-text fallback');
    return freeTextAnswerFallback(
      request.message,
      `project=${projectId} chapter=${request.context.chapter_id || 'none'} page=${request.context.route || '?'}`
    );
  }
}
